File: src/main/java/by/masarnovsky/releasedemon/entity/Artist.java
package by.masarnovsky.releasedemon.entity;

import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import javax.persistence.*;
import java.util.HashSet;
import java.util.Set;

@Entity
@Getter
@Setter
@NoArgsConstructor
public class Artist {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    private String name;

    private String mbid;

    private String spotifyId;

    @OneToMany(mappedBy = "artist")
    private Set<Album> albums = new HashSet<>();

    @ManyToMany(mappedBy = "artists")
    private Set<User> users = new HashSet<>();

    public void addAlbum(Album album) {
        album.setArtist(this);
        albums.add(album);
    }
}

File: src/main/kotlin/by/masarnovsky/releasedemon/entity/Artist.kt
package by.masarnovsky.releasedemon.entity

import javax.persistence.*

@Entity
class Artist(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Int?,
    val name: String,
    var mbid: String?,
    var spotifyId: String?,

    @OneToMany(mappedBy = "artist")
    var albums: MutableSet<Album>,

    @ManyToMany(mappedBy = "artists")
    var users: MutableSet<User>,
) {

    fun addAlbum(album: Album) {
        album.artist = this
        albums.add(album)
    }

    fun toDTO(): ArtistDTO {
        return ArtistDTO(this.id, this.name, this.mbid, this.albums.map { it.toDTO() })
    }
}

data class ArtistDTO(
    var id: Int?,
    val name: String,
    var mbid: String?,
    val albums: List<AlbumDTO>
)